File: src/main/java/uk/co/lukestevens/geoguessr/models/Game.java
package uk.co.lukestevens.geoguessr.models;

import uk.co.lukestevens.utils.Dates;

import javax.persistence.*;
import java.time.Instant;
import java.util.ArrayList;
import java.util.Date;
import java.util.List;

@Entity
@Table(name = "games", schema = "geoguessr")
public class Game {

    @Id
    @GeneratedValue(strategy=GenerationType.IDENTITY)
    @Column(name = "game_id")
    private Long id;

    @Column(name = "challenge_token")
    private String challengeToken;

    @Column(name = "created_at")
    private Instant createdAt;

    @Column(name = "results_posted")
    private boolean resultsPosted;

    @Column(name = "post_results_after")
    private Instant postResultsAfter;

    @ManyToOne(fetch = FetchType.EAGER)
    @JoinColumn(name = "game_option_id")
    private GameOption gameOption;

    @OneToMany(mappedBy = "game", cascade = CascadeType.ALL, fetch = FetchType.EAGER)
    private List<PlayerScore> playerScores = new ArrayList<>();

    @ManyToOne(fetch = FetchType.EAGER)
    @JoinColumn(name = "league_id")
    private League league;

    public Game(String challengeToken, GameOption gameOption) {
        this.challengeToken = challengeToken;
        this.gameOption = gameOption;
        this.createdAt = Dates.now().toInstant();
    }

    public Game() { }

    public Long getId() {
        return id;
    }

    public String getChallengeToken() {
        return challengeToken;
    }

    public Instant getCreatedAt() {
        return createdAt;
    }

    public void setResultsPosted(boolean resultsPosted) {
        this.resultsPosted = resultsPosted;
    }

    public boolean isResultsPosted() {
        return resultsPosted;
    }

    public GameOption getGameOption() {
        return gameOption;
    }

    public List<PlayerScore> getPlayerScores() {
        return playerScores;
    }

    public League getLeague() {
        return league;
    }

    public void setLeague(League league) {
        this.league = league;
    }

    public void setPostResultsAfter(Instant postResultsAfter) {
        this.postResultsAfter = postResultsAfter;
    }

    public void setId(Long id) {
        this.id = id;
    }

    public void setChallengeToken(String challengeToken) {
        this.challengeToken = challengeToken;
    }

    public void setCreatedAt(Instant createdAt) {
        this.createdAt = createdAt;
    }

    public Instant getPostResultsAfter() {
        return postResultsAfter;
    }

    public void setGameOption(GameOption gameOption) {
        this.gameOption = gameOption;
    }

    public void setPlayerScores(List<PlayerScore> playerScores) {
        this.playerScores = playerScores;
    }
}

File: src/main/kotlin/uk/co/lukestevens/geoguessr/models/Game.kt
package uk.co.lukestevens.geoguessr.models

import uk.co.lukestevens.utils.Dates
import java.time.Instant
import javax.persistence.CascadeType
import javax.persistence.Column
import javax.persistence.Entity
import javax.persistence.FetchType
import javax.persistence.GeneratedValue
import javax.persistence.GenerationType
import javax.persistence.Id
import javax.persistence.JoinColumn
import javax.persistence.ManyToOne
import javax.persistence.OneToMany
import javax.persistence.Table

@Entity
@Table(name = "games", schema = "geoguessr")
class Game {
    @Id
    @GeneratedValue(strategy = GenerationType.IDENTITY)
    @Column(name = "game_id")
    var id: Long? = null

    @Column(name = "challenge_token")
    var challengeToken: String? = null

    @Column(name = "created_at")
    var createdAt: Instant = Dates.now().toInstant()

    @Column(name = "results_posted")
    var isResultsPosted = false

    @Column(name = "post_results_after")
    var postResultsAfter: Instant? = null

    @ManyToOne(fetch = FetchType.EAGER)
    @JoinColumn(name = "game_option_id")
    var gameOption: GameOption? = null

    @OneToMany(mappedBy = "game", cascade = [CascadeType.ALL], fetch = FetchType.EAGER)
    var playerScores: MutableList<PlayerScore> = ArrayList()

    @ManyToOne(fetch = FetchType.EAGER)
    @JoinColumn(name = "league_id")
    var league: League? = null

    constructor(challengeToken: String, gameOption: GameOption) {
        this.challengeToken = challengeToken
        this.gameOption = gameOption
        createdAt = Dates.now().toInstant()
    }

    constructor()
}